### ISSUE TITLE
Check if page already exists before calling createPage to avoid nondeterministic routing

### ISSUE THREAD

**Issue by vai0** [gatsbyjs/gatsby#15866]
We currently have a set of pages in  `/src/pages/` and another set of pages built programatically with `createPage` in the `createPages` gatsby node API hook. 

We want to prevent programmatically created pages from potentially overlapping pages in `/src/pages`. 

I've tried querying for all pages in the `createPages` hook with
```
allSitePage {
    nodes {
        path
    }
}
``` 
but it seems like pages from `/src/pages/` are sourced AFTER the `createPages` hook as that query is only returning `['dev-404-page']` and I can't add a check to see if the page already exists in `/src/pages` before calling `createPage`.

In conclusion -- how do we programmatically create pages and ensure they do not overlap/overwrite pages in `/src/pages/`?

**Comment by freiksenet** [comment]
Hi! 

We recommend that in this case you create all pages using `createPages` API. You can put files in `src/pages` to eg `src/templates/pages` and then go through all files in that directory in your `createPages`. This way you'll have a list of pages that you can check against.

Hope that helps!

**Comment by vai0** [comment]
Appreciate the quick response! I would prefer to keep the default directory structure so there's less potential confusion for future contributors to our project.

In the interim we're wrapping all createPage in a createPageSafely wrapper, including the existing `createPage`s I already have in `onCreatePage` and keep track of built pages this way instead.  

So something like this:
```
const createdPages = [];
function createPageSafely(createPage, { path, component, context }) {
    const cleanPath = _.trim(path, "/");

    if (createdPages.includes(cleanPath)) {
        throw Error(
            `Attempted to create page "/${cleanPath}", but page "/${cleanPath}" already exists. Exited to prevent potential non-deterministic routing behavior.`
        );
    }
    createPage({
        path,
        component,
        context,
    });
    createdPages.push(cleanPath);
}
```

**Comment by freiksenet** [comment]
@vai0 Did you test if it works? IIRC `page-creator` runs last, which means you won't have that array populated by built-in pages. But I might be mistaken.

**Comment by vai0** [comment]
@freiksenet Yep tested and it works.`onCreatePage` hook is called for built-in pages - and the `createdPages` array gets filled with built-in page paths.
